llvm/include/llvm/Support/PropertySetIO.h

```diff
@@ -0,0 +1,164 @@
+//   '['<PropertyValue set name>']'
+//   <property name>=<property type>'|'<property value>
+// where
+//   <PropertyValue set name>, <property name> are strings
+//   <property type> - string representation of the property type
+//   <property value> - string representation of the property value.
+//
+// For example:
+// [Staff/Ages]
+// person1=1|20
+// person2=1|25
+// [Staff/Experience]
+// person1=1|1
+// person2=1|2
+// person3=1|12
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <istream>
+#include <map>
+#include <memory>
+#include <string>
+
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/StringRef.h"
+
+#include "llvm/Support/Error.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+
+namespace llvm {
+namespace util {
+
+// Represents a property value. PropertyValue name is stored in the encompassing
+// container.
+class PropertyValue {
+public:
+  // Defines supported property types
+  enum Type { first = 0, NONE = first, UINT32, last = UINT32 };
+
+  // Translates C++ type to the corresponding type tag.
+  template <typename T> static Type getTypeTag();
+
+  // Casts from int value to a type tag.
+  static Expected<Type> getTypeTag(int T) {
+    if (T < first || T > last)
+      return createStringError(std::error_code(), "bad property type " + T);
+    return static_cast<Type>(T);
+  }
+
+  PropertyValue() : Ty(NONE) {}
+  PropertyValue(Type T) : Ty(T) {}
+
+  PropertyValue(uint32_t Val) : Ty(UINT32), Val({Val}) {}
+  PropertyValue(const PropertyValue &P) : Ty(P.Ty), Val({P.Val}) {}
+  PropertyValue(PropertyValue &&P) : Ty(P.Ty), Val({P.Val}) {}
+
+  PropertyValue &operator=(PropertyValue &&P) = default;
+
+  PropertyValue &operator=(const PropertyValue &P) = default;
+
+  // get property value as unsigned 32-bit integer
+  uint32_t asUint32() const {
+    assert(Ty == UINT32);
+    return Val.UInt32Val;
+  }
+
+  bool isValid() const { return getType() != NONE; }
+
+  // set property value; the 'T' type must be convertible to a property type tag
+  template <typename T> void set(T V) {
+    assert(getTypeTag<T>() == Ty);
+    getValueRef<T>() = V;
+  }
+
+  Type getType() const { return Ty; }
+
+  size_t size() const {
+    switch (Ty) {
+    case UINT32:
+      return sizeof(Val.UInt32Val);
+    default:
+      llvm_unreachable_internal("unsupported property type");
+    }
+  }
+
+private:
+  template <typename T> T &getValueRef();
+
+  Type Ty;
+  union {
+    uint32_t UInt32Val;
+  } Val;
+};
+
+std::ostream &operator<<(std::ostream &Out, const PropertyValue &V);
+
+// A property set. Preserves insertion order when iterating elements.
+using PropertySet = MapVector<StringRef, PropertyValue>;
+
+// A "registry" of multiple property sets. Maps a property set name to its
+// contents. Can be read/written.
+class PropertySetRegistry {
+public:
+  // using MapTy = std::map<StringRef, PropertySet>;
```

Comment:
Comment is still there.

Reply:
Somehow I did not check it in. Should be fixed now. @sndmitriev, can you please re-approve